File: backend/service/admin/biz/comment_info.go
package biz

import (
	"context"
	"strings"

	adminv1 "shop/api/gen/go/admin/v1"
	"shop/pkg/biz"
	_const "shop/pkg/const"
	"shop/pkg/errorsx"
	"shop/pkg/gen/data"
	"shop/pkg/gen/models"
	"shop/pkg/queue"

	"github.com/liujitcn/go-utils/mapper"
	_string "github.com/liujitcn/go-utils/string"
	"github.com/liujitcn/gorm-kit/repository"
	"gorm.io/gen"
)

// CommentInfoCase 评论管理业务实例。
type CommentInfoCase struct {
	*biz.BaseCase
	*data.CommentInfoRepository
	tx                    data.Transaction
	commentTagCase        *CommentTagCase
	commentDiscussionCase *CommentDiscussionCase
	commentAiCase         *CommentAiCase
	commentReviewCase     *CommentReviewCase
	baseUserCase          *BaseUserCase
	commentInfoMapper     *mapper.CopierMapper[adminv1.CommentInfo, models.CommentInfo]
}

// NewCommentInfoCase 创建评论管理业务实例。
func NewCommentInfoCase(
	baseCase *biz.BaseCase,
	commentInfoRepo *data.CommentInfoRepository,
	tx data.Transaction,
	commentTagCase *CommentTagCase,
	commentDiscussionCase *CommentDiscussionCase,
	commentAiCase *CommentAiCase,
	commentReviewCase *CommentReviewCase,
	baseUserCase *BaseUserCase,
) *CommentInfoCase {
	commentInfoMapper := mapper.NewCopierMapper[adminv1.CommentInfo, models.CommentInfo]()
	commentInfoMapper.AppendConverters(mapper.NewJSONTypeConverter[[]int64]().NewConverterPair())
	return &CommentInfoCase{
		BaseCase:              baseCase,
		CommentInfoRepository: commentInfoRepo,
		tx:                    tx,
		commentTagCase:        commentTagCase,
		commentDiscussionCase: commentDiscussionCase,
		commentAiCase:         commentAiCase,
		commentReviewCase:     commentReviewCase,
		baseUserCase:          baseUserCase,
		commentInfoMapper:     commentInfoMapper,
	}
}

// PageCommentInfos 分页查询评论审核列表。
func (c *CommentInfoCase) PageCommentInfos(ctx context.Context, req *adminv1.PageCommentInfosRequest) (*adminv1.PageCommentInfosResponse, error) {
	query := c.Query(ctx).CommentInfo
	opts := make([]repository.QueryOption, 0, 7)
	opts = append(opts, repository.Order(query.CreatedAt.Desc()))
	if req.GoodsId != nil && req.GetGoodsId() > 0 {
		opts = append(opts, repository.Where(query.GoodsID.Eq(req.GetGoodsId())))
	}
	// 传入商品名关键字时，按商品名称快照模糊匹配。
	if strings.TrimSpace(req.GetGoodsName()) != "" {
		opts = append(opts, repository.Where(query.GoodsNameSnapshot.Like("%"+strings.TrimSpace(req.GetGoodsName())+"%")))
	}
	// 传入用户昵称关键字时，按用户昵称快照模糊匹配。
	if strings.TrimSpace(req.GetUserName()) != "" {
		opts = append(opts, repository.Where(query.UserNameSnapshot.Like("%"+strings.TrimSpace(req.GetUserName())+"%")))
	}
	if req.GoodsScore != nil && req.GetGoodsScore() > 0 {
		opts = append(opts, repository.Where(query.GoodsScore.Eq(req.GetGoodsScore())))
	}
	if req.MinGoodsScore != nil && req.GetMinGoodsScore() > 0 {
		opts = append(opts, repository.Where(query.GoodsScore.Gte(req.GetMinGoodsScore())))
	}
	if req.MaxGoodsScore != nil && req.GetMaxGoodsScore() > 0 {
		opts = append(opts, repository.Where(query.GoodsScore.Lte(req.GetMaxGoodsScore())))
	}
	if req.HasPendingDiscussion != nil && req.GetHasPendingDiscussion() {
		opts = append(opts, repository.Where(query.PendingDiscussionCount.Gt(0)))
	}
	if req.Status != nil {
		opts = append(opts, repository.Where(query.Status.Eq(int32(req.GetStatus()))))
	}

	list, total, err := c.Page(ctx, req.GetPageNum(), req.GetPageSize(), opts...)
	if err != nil {
		return nil, err
	}

	resList := make([]*adminv1.CommentInfo, 0, len(list))
	for _, item := range list {
		resList = append(resList, c.commentInfoMapper.ToDTO(item))
	}
	return &adminv1.PageCommentInfosResponse{CommentInfos: resList, Total: int32(total)}, nil
}

// GetGoodsCommentInfo 按商品查询评论聚合信息。
func (c *CommentInfoCase) GetGoodsCommentInfo(ctx context.Context, goodsID int64) (*adminv1.GoodsCommentInfoResponse, error) {
	// 商品编号为空时，无法定位评论聚合范围。
	if goodsID <= 0 {
		return nil, errorsx.InvalidArgument("商品ID不能为空")
	}

	query := c.Query(ctx).CommentInfo
	opts := make([]repository.QueryOption, 0, 2)
	opts = append(opts, repository.Where(query.GoodsID.Eq(goodsID)))
	opts = append(opts, repository.Order(query.CreatedAt.Desc()))
	commentList, err := c.List(ctx, opts...)
	if err != nil {
		return nil, err
	}

	commentIDs := make([]int64, 0, len(commentList))
	resCommentList := make([]*adminv1.CommentInfo, 0, len(commentList))
	for _, item := range commentList {
		commentIDs = append(commentIDs, item.ID)
		resCommentList = append(resCommentList, c.commentInfoMapper.ToDTO(item))
	}

	var tagList []*adminv1.CommentTag
	tagList, err = c.commentTagCase.ListByGoodsID(ctx, goodsID)
	if err != nil {
		return nil, err
	}

	var discussionList []*adminv1.CommentDiscussion
	discussionList, err = c.commentDiscussionCase.ListByCommentIDs(ctx, commentIDs)
	if err != nil {
		return nil, err
	}

	var aiList []*adminv1.CommentAi
	aiList, err = c.commentAiCase.ListByGoodsID(ctx, goodsID)
	if err != nil {
		return nil, err
	}

	return &adminv1.GoodsCommentInfoResponse{
		CommentInfos:       resCommentList,
		CommentTags:        tagList,
		CommentDiscussions: discussionList,
		CommentAis:         aiList,
	}, nil
}

// GetCommentInfo 查询评论详情。
func (c *CommentInfoCase) GetCommentInfo(ctx context.Context, commentID int64) (*adminv1.CommentInfoDetail, error) {
	commentInfo, err := c.FindByID(ctx, commentID)
	if err != nil {
		return nil, err
	}

	var tagList []*adminv1.CommentTag
	tagList, err = c.commentTagCase.ListByGoodsID(ctx, commentInfo.GoodsID)
	if err != nil {
		return nil, err
	}

	var discussionList []*adminv1.CommentDiscussion
	discussionList, err = c.commentDiscussionCase.ListByCommentIDs(ctx, []int64{commentInfo.ID})
	if err != nil {
		return nil, err
	}

	var aiList []*adminv1.CommentAi
	aiList, err = c.commentAiCase.ListByGoodsID(ctx, commentInfo.GoodsID)
	if err != nil {
		return nil, err
	}

	var reviewList []*adminv1.CommentReview
	reviewList, err = c.commentReviewCase.ListByTarget(ctx, _const.COMMENT_REVIEW_TARGET_TYPE_COMMENT, commentInfo.ID)
	if err != nil {
		return nil, err
	}

	return &adminv1.CommentInfoDetail{
		Comment:            c.commentInfoMapper.ToDTO(commentInfo),
		CommentTags:        tagList,
		CommentDiscussions: discussionList,
		CommentAis:         aiList,
		CommentReviews:     reviewList,
	}, nil
}

// SetCommentInfoStatus 设置评论审核状态。
func (c *CommentInfoCase) SetCommentInfoStatus(ctx context.Context, req *adminv1.SetCommentInfoStatusRequest) error {
	err := validateCommentStatus(int32(req.GetStatus()))
	if err != nil {
		return err
	}
	authInfo, err := c.GetAuthInfo(ctx)
	if err != nil {
		return err
	}

	commentInfo, err := c.FindByID(ctx, req.GetId())
	if err != nil {
		return err
	}
	operatorName := c.operatorName(ctx, authInfo.UserId)
	err = c.tx.Transaction(ctx, func(txCtx context.Context) error {
		// 评论从通过改为不通过时，释放已计入筛选项的标签提及次数并清空命中标签。
		if commentInfo.Status == _const.COMMENT_STATUS_APPROVED && int32(req.GetStatus()) != _const.COMMENT_STATUS_APPROVED {
			tagIDs := _string.ConvertJsonStringToInt64Array(commentInfo.TagID)
			updateErr := c.commentTagCase.DecreaseMentionCount(txCtx, tagIDs)
			if updateErr != nil {
				return updateErr
			}
			updateErr = c.updateTagIDs(txCtx, req.GetId(), []int64{})
			if updateErr != nil {
				return updateErr
			}
		}
		updateErr := c.updateStatus(txCtx, req.GetId(), int32(req.GetStatus()))
		if updateErr != nil {
			return updateErr
		}
		return c.commentReviewCase.CreateReview(txCtx, &models.CommentReview{
			TargetType:   _const.COMMENT_REVIEW_TARGET_TYPE_COMMENT,
			TargetID:     req.GetId(),
			Type:         _const.COMMENT_REVIEW_TYPE_MANUAL,
			Status:       commentReviewStatusByCommentStatus(int32(req.GetStatus())),
			Tags:         _string.ConvertAnyToJsonString([]string{}),
			OperatorID:   authInfo.UserId,
			OperatorName: operatorName,
			Reason:       strings.TrimSpace(req.GetReason()),
		})
	})
	if err != nil {
		return err
	}
	// 评论进入或离开通过态后异步刷新商品评价摘要，不阻塞后台审核操作。
	if int32(req.GetStatus()) == _const.COMMENT_STATUS_APPROVED || commentInfo.Status == _const.COMMENT_STATUS_APPROVED {
		queue.DispatchCommentAiRefresh(commentInfo.GoodsID)
	}
	return nil
}

// updateTagIDs 更新评论命中的标签编号。
func (c *CommentInfoCase) updateTagIDs(ctx context.Context, commentID int64, tagIDs []int64) error {
	query := c.Query(ctx).CommentInfo
	result, err := query.WithContext(ctx).
		Where(query.ID.Eq(commentID)).
		Update(query.TagID, _string.ConvertAnyToJsonString(tagIDs))
	if err != nil {
		return err
	}
	if result.RowsAffected == 0 {
		return errorsx.ResourceNotFound("评论不存在")
	}
	return nil
}

// updateStatus 更新评论审核状态。
func (c *CommentInfoCase) updateStatus(ctx context.Context, commentID int64, status int32) error {
	query := c.Query(ctx).CommentInfo
	result, err := query.WithContext(ctx).
		Where(query.ID.Eq(commentID)).
		Update(query.Status, status)
	if err != nil {
		return err
	}
	if result.RowsAffected == 0 {
		return errorsx.ResourceNotFound("评论不存在")
	}
	return nil
}

// changeDiscussionCount 调整评论讨论缓存数量。
func (c *CommentInfoCase) changeDiscussionCount(ctx context.Context, commentID int64, status int32, delta int32) error {
	if delta == 0 {
		return nil
	}
	query := c.Query(ctx).CommentInfo
	update := query.DiscussionCount.Add(delta)
	conditions := []gen.Condition{query.ID.Eq(commentID)}
	switch status {
	case _const.COMMENT_STATUS_PENDING_REVIEW:
		update = query.PendingDiscussionCount.Add(delta)
		if delta < 0 {
			conditions = append(conditions, query.PendingDiscussionCount.Gt(0))
		}
	case _const.COMMENT_STATUS_APPROVED:
		update = query.DiscussionCount.Add(delta)
		if delta < 0 {
			conditions = append(conditions, query.DiscussionCount.Gt(0))
		}
	default:
		return nil
	}
	_, err := query.WithContext(ctx).Where(conditions...).UpdateSimple(update)
	return err
}

// operatorName 查询后台操作人展示名称。
func (c *CommentInfoCase) operatorName(ctx context.Context, userID int64) string {
	baseUser, err := c.baseUserCase.FindByID(ctx, userID)
	if err != nil || baseUser == nil {
		return "管理员"
	}
	operatorName := strings.TrimSpace(baseUser.NickName)
	// 后台用户昵称为空时，回退到账号名作为操作人名称。
	if operatorName == "" {
		operatorName = strings.TrimSpace(baseUser.UserName)
	}
	if operatorName == "" {
		return "管理员"
	}
	return operatorName
}

File: frontend/admin/src/api/admin/comment_info.ts
import service from "@/utils/request";
import {
  type CommentAi,
  type CommentDiscussion,
  type CommentInfo,
  type CommentInfoDetail,
  type CommentInfoService,
  type CommentReview,
  type CommentTag,
  type GetCommentInfoRequest,
  type GetGoodsCommentInfoRequest,
  type GoodsCommentInfoResponse,
  type ListCommentReviewsRequest,
  type ListCommentReviewsResponse,
  type PageCommentDiscussionsRequest,
  type PageCommentDiscussionsResponse,
  type PageCommentInfosRequest,
  type PageCommentInfosResponse,
  type SetCommentDiscussionStatusRequest,
  type SetCommentInfoStatusRequest
} from "@/rpc/admin/v1/comment_info";
import type { Empty } from "@/rpc/google/protobuf/empty";

const COMMENT_INFO_URL = "/v1/admin/comment/info";

type PageCommentInfosHTTPResponse = Partial<PageCommentInfosResponse> & {
  /** 旧协议评论分页数据字段 */
  list?: CommentInfo[];
};

type CommentInfoDetailHTTPResponse = Partial<CommentInfoDetail> & {
  /** 旧协议商品评论标签字段 */
  tagList?: CommentTag[];
  /** 旧协议评论讨论列表字段 */
  discussionList?: CommentDiscussion[];
  /** 旧协议商品评论 AI 摘要字段 */
  aiList?: CommentAi[];
  /** 旧协议评论审核记录字段 */
  reviewList?: CommentReview[];
};

type GoodsCommentInfoHTTPResponse = Partial<GoodsCommentInfoResponse> & {
  /** 旧协议评论列表字段 */
  commentList?: CommentInfo[];
  /** 旧协议商品评论标签字段 */
  tagList?: CommentTag[];
  /** 旧协议评论讨论列表字段 */
  discussionList?: CommentDiscussion[];
  /** 旧协议商品评论 AI 摘要字段 */
  aiList?: CommentAi[];
};

type PageCommentDiscussionsHTTPResponse = Partial<PageCommentDiscussionsResponse> & {
  /** 旧协议讨论分页数据字段 */
  list?: CommentDiscussion[];
};

type ListCommentReviewsHTTPResponse = Partial<ListCommentReviewsResponse> & {
  /** 旧协议审核记录字段 */
  list?: CommentReview[];
};

/** Admin评论管理服务。 */
export class CommentInfoServiceImpl implements CommentInfoService {
  /** 查询评论分页列表。 */
  async PageCommentInfos(request: PageCommentInfosRequest): Promise<PageCommentInfosResponse> {
    const response = await service<PageCommentInfosRequest, PageCommentInfosHTTPResponse>({
      url: `${COMMENT_INFO_URL}`,
      method: "get",
      params: request
    });
    // 兼容未统一生成前的旧响应 list，同时向新协议 comment_infos 收敛。
    const comment_infos = response.comment_infos ?? response.list ?? [];
    return { ...response, comment_infos, total: response.total ?? 0 } as PageCommentInfosResponse;
  }

  /** 按商品查询评论聚合信息。 */
  async GetGoodsCommentInfo(request: GetGoodsCommentInfoRequest): Promise<GoodsCommentInfoResponse> {
    const response = await service<GetGoodsCommentInfoRequest, GoodsCommentInfoHTTPResponse>({
      url: `${COMMENT_INFO_URL}/goods/${request.goods_id}`,
      method: "get"
    });
    // 兼容旧聚合字段，同时向新协议复数资源字段收敛。
    const comment_infos = response.comment_infos ?? response.commentList ?? [];
    const comment_tags = response.comment_tags ?? response.tagList ?? [];
    const comment_discussions = response.comment_discussions ?? response.discussionList ?? [];
    const comment_ais = response.comment_ais ?? response.aiList ?? [];
    return { ...response, comment_infos, comment_tags, comment_discussions, comment_ais } as GoodsCommentInfoResponse;
  }

  /** 查询评论详情。 */
  async GetCommentInfo(request: GetCommentInfoRequest): Promise<CommentInfoDetail> {
    const response = await service<GetCommentInfoRequest, CommentInfoDetailHTTPResponse>({
      url: `${COMMENT_INFO_URL}/${request.id}`,
      method: "get"
    });
    // 兼容旧详情字段，同时向新协议复数资源字段收敛。
    const comment_tags = response.comment_tags ?? response.tagList ?? [];
    const comment_discussions = response.comment_discussions ?? response.discussionList ?? [];
    const comment_ais = response.comment_ais ?? response.aiList ?? [];
    const comment_reviews = response.comment_reviews ?? response.reviewList ?? [];
    return {
      ...response,
      comment: response.comment,
      comment_tags,
      comment_discussions,
      comment_ais,
      comment_reviews
    } as CommentInfoDetail;
  }

  /** 查询评论审核记录列表。 */
  async ListCommentReviews(request: ListCommentReviewsRequest): Promise<ListCommentReviewsResponse> {
    const response = await service<ListCommentReviewsRequest, ListCommentReviewsHTTPResponse>({
      url: `${COMMENT_INFO_URL}/review`,
      method: "get",
      params: request
    });
    const comment_reviews = response.comment_reviews ?? response.list ?? [];
    return { ...response, comment_reviews } as ListCommentReviewsResponse;
  }

  /** 设置评论审核状态。 */
  SetCommentInfoStatus(request: SetCommentInfoStatusRequest): Promise<Empty> {
    return service<SetCommentInfoStatusRequest, Empty>({
      url: `${COMMENT_INFO_URL}/${request.id}/status`,
      method: "put",
      data: { ...request, reason: request.reason ?? "" }
    });
  }

  /** 查询评论讨论分页列表。 */
  async PageCommentDiscussions(request: PageCommentDiscussionsRequest): Promise<PageCommentDiscussionsResponse> {
    const response = await service<PageCommentDiscussionsRequest, PageCommentDiscussionsHTTPResponse>({
      url: `${COMMENT_INFO_URL}/${request.comment_id}/discussion`,
      method: "get",
      params: request
    });
    // 兼容未统一生成前的旧响应 list，同时向新协议 comment_discussions 收敛。
    const comment_discussions = response.comment_discussions ?? response.list ?? [];
    return { ...response, comment_discussions, total: response.total ?? 0 } as PageCommentDiscussionsResponse;
  }

  /** 设置评论讨论审核状态。 */
  SetCommentDiscussionStatus(request: SetCommentDiscussionStatusRequest): Promise<Empty> {
    return service<SetCommentDiscussionStatusRequest, Empty>({
      url: `${COMMENT_INFO_URL}/discussion/${request.id}/status`,
      method: "put",
      data: { ...request, reason: request.reason ?? "" }
    });
  }
}

export const defCommentInfoService = new CommentInfoServiceImpl();
